feat: support for relation data schema as Python classes (#1701)
canonical/operator: docs/howto/manage-libraries.md

    assert isinstance(custom_event, DatabaseReadyEvent)
    assert custom_event.credential_secret.id == secret.id
```

### Write a library that manages a relation interface

First, for a new interface, design the interface and the process related charms
will use to populate the relation data. In the simplest case, the providing
charm might populate the local app data when the relation is created, but a
conversation between the charms is more common, where the requiring charm will
populate its local data with a request, and the providing charm will use that to
provide a suitable response. In more complex cases, this conversation might have
multiple stages.

> See more: {ref}`manage-interfaces`

> If the library is implementing an existing interface, find the schema and
> details of the processes by which relation data is added on Charmhub in the
> integrations tabs, or by navigating to `https://charmhub.io/integrations/{integration-name}`.
> Alternatively, the interface details can be found in the
> [charm-relation-interfaces](https://github.com/canonical/charm-relation-interfaces)
> repository.
> 
> In the documentation for the interface, find the description of the various

Reviewer: Following the trail through Charmhub, I'm not totally sure what "the processes by which relation data is added" refers to. Maybe this would be fine (also including a couple more small suggestions):

```suggestion
If the library is implementing an existing interface, find the interface documentation by following links from the Integrations tab on Charmhub, or by navigating to `https://charmhub.io/integrations/{integration-name}`.
Alternatively, the interface documentation can be found in the
[charm-relation-interfaces](https://github.com/canonical/charm-relation-interfaces)
repository.

In the interface documentation, find the description of the various
```

    )
    type: TransportProtocolType = pydantic.Field(
        description="The transport protocol used by this receiver.",
        examples=["http", "grpc"],
    )

class Receiver(pydantic.BaseModel):
    """Specification of an active receiver."""
    protocol: ProtocolType = pydantic.Field(description="Receiver protocol name and type.")
    url: str = pydantic.Field(
        description="""URL at which the receiver is reachable. If there's an ingress, it would be the external URL.
        Otherwise, it would be the service's fqdn or internal IP.
        If the protocol type is grpc, the url will not contain a scheme.""",
        examples=[
            "http://traefik_address:2331",
            "https://traefik_address:2331",
            "http://tempo_public_ip:2331",
            "https://tempo_public_ip:2331",
            "tempo_public_ip:2331",
        ],
    )

class TracingProviderAppData(pydantic.BaseModel):
    receivers: list[Receiver] = pydantic.Field(
        description="A list of enabled receivers in the form of the protocol they use and their resolvable server url.",
    )
```

```{tip}

> Ops defaults to transparently serialising each field to and from JSON, so you
> do not need to wrap each field in `pydantic.Json`.

Reviewer: Is the point here that all data in databags is stored as JSON? At first I was slightly confused by this tip, because JSON hadn't been mentioned up until this point. I'm also wondering what "defaults to" means - would I be able to disable it if I wanted? How about something like this (if correct):

```suggestion
Internally, Juju stores databag fields as JSON strings. Ops automatically serialises each field to and from JSON, so you
do not need to wrap each field in `pydantic.Json`.
```

Author: > Is the point here that all data in databags is stored as JSON?

Yes. In particular, that at the moment the schemas (as found in libraries and charm-relation-interfaces) tend to look like:

```python
class TracingRequirerData(BaseModel):
    receivers: Json[List[str]] = Field(
        ..., description="List of protocols that the requirer wishes to use."
    )
```

But ops is taking care of the JSON serialisation and deserialisation so there's no need for the `Json` that's there.

> I'm also wondering what "defaults to" means - would I be able to disable it if I wanted?

Yes, `load` takes an `decoder` and `save` takes an `encoder` argument, and they default to `json.loads` and `json.dumps` but can be anything else you want.

> Internally, Juju stores databag fields as JSON strings.

So this isn't correct. Juju just has str:str and knows nothing at all about what that string might be. It's Ops that's defaulting to encoding/decoding each field to/from JSON.

What about something like this?

```suggestion
The Ops [](ops.Relation.load) and [](ops.Relation.save) methods serialise and deserialise the values of each field, and default to using JSON, so you do not need to wrap each field in `pydantic.Json`.
```

Or perhaps it's not needed at all? It's really mainly there for those people that are migrating from existing usage. Maybe it should be in a migration tip instead or something like that?

```

> In your `src/charm.py`, use the class you created to load and save data from the
> relation.

Reviewer: ```suggestion
In your `src/charm.py` file, use the class you created to load and save data from the
relation:
```
